lib/ui/window.dart

```diff
@@ -145,12 +146,46 @@ class Locale {
   /// The primary language subtag must not be null. The region subtag is
   /// optional.
   ///
-  /// The values are _case sensitive_, and should match the case of the relevant
-  /// subtags in the [IANA Language Subtag
-  /// Registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry).
-  /// Typically this means the primary language subtag should be lowercase and
-  /// the region subtag should be uppercase.
-  const Locale(this._languageCode, [ this._countryCode ]) : assert(_languageCode != null), assert(_languageCode != '');
+  /// The subtag values are _case sensitive_ and should be one of the valid
+  /// subtags according to CLDR supplemental data:
+  /// [language](http://unicode.org/cldr/latest/common/validity/language.xml),
+  /// [region](http://unicode.org/cldr/latest/common/validity/region.xml). The
+  /// primary language subtag should be two to three lowercase letters. The
+  /// region region subtag should be two uppercase letters or three digits. See
+  /// the [Unicode Language
+  /// Identifier](https://www.unicode.org/reports/tr35/#Unicode_language_identifier)
+  /// specification.
+  ///
+  /// Validity is not checked by default, but some methods may throw away
+  /// invalid data.
+  ///
+  /// See also: [Locale.create].
+  const Locale(
+    this._languageCode, [
+    this._countryCode,
+  ]) : assert(_languageCode != null),
+       assert(_languageCode != ''),
+       scriptCode = null;
+
+  /// Creates a new Locale object.
+  ///
+  /// The keyword arguments specify the subtags of the Locale.
+  ///
+  /// The subtag values are _case sensitive_ and should be valid subtags
+  /// according to CLDR supplemental data:
+  /// [language](http://unicode.org/cldr/latest/common/validity/language.xml),
+  /// [script](http://unicode.org/cldr/latest/common/validity/script.xml),
+  /// [region](http://unicode.org/cldr/latest/common/validity/region.xml).
+  /// for each of languageCode, scriptCode and countryCode respectively.
+  ///
+  /// Validity is not checked by default, but some methods may throw away
+  /// invalid data.
+  const Locale.fromSubtags({
+    String languageCode = 'und',
+    this.scriptCode,
+    String countryCode,
+  }) : _languageCode = languageCode,
```

Comment:
we should probably assert that languageCode is not null (and mention in the constructor docs)

Reply:
Adding an assert that it also isn't '', to match the unnamed constructor. However zero-length string could also be argued to be just a particular version of the previously proposed length checks. :)

Reply:
Constructor docs already state they should be valid subtags, which implies exclusion of zero-length strings and null?

Reply:
let's change "should" to "must" in the docs to be clearer about this.

The check for empty string is a special case in that I think it'd be common to pass '' where null is meant.

Reply:
(to put it another way, even given bogus input, the empty string is never allowed. one could imagine being given a bogus language code that is too long, or uses the wrong characters, or whatever, and short of implementing a validator there's no good response. But it's a structural failure if you're given the empty string, not just bad data.)

Reply:
Of course I'd love to treat other invalid subtag lengths as "special case invalid" too, given its a structural failure to pass a Script tag as a Language tag or a Region tag, and vice versa - and Script subtags are the only Language Identifier subtag that should be length 4...

That said: I've updated the docs. Missed this suggestion due to "outdated" comments, I'm still wanting a better way to see and deal with these, seems it might be lacking.